File: src/Unfold.Core/PixelDocument.cs
namespace Unfold.Core;

public enum PixelTool { Pencil, Eraser, Fill, Eyedropper, Line, Rectangle, Ellipse }
public readonly record struct PixelPoint(int X, int Y);

public sealed class PixelLayer
{
    public string Name { get; set; } = "Layer 1";
    public double Opacity { get; set; } = 1;
    public List<uint[]> Frames { get; set; } = [];
    public PixelLayer Clone() => new() { Name = Name, Opacity = Opacity, Frames = Frames.Select(f => (uint[])f.Clone()).ToList() };
}

/// <summary>Top-left row-major, straight alpha 0xAARRGGBB pixels. No UI dependency.</summary>
public sealed class PixelDocument
{
    public const int MaxSide = 128, MaxFrames = 24, MaxLayers = 16;
    public int Width { get; private set; }
    public int Height { get; private set; }
    public double Fps { get; set; } = 12;
    public string Name { get; set; } = "Unfold Character";
    public string Description { get; set; } = "";
    public List<PixelLayer> Layers { get; set; }
    public int FrameCount => Layers[0].Frames.Count;
    public long ByteCount => (long)Width * Height * FrameCount * Layers.Count * 4;

    public PixelDocument(int width = 64, int height = 64)
    {
        CheckSize(width, height);
        Width = width; Height = height;
        Layers = [new() { Frames = [new uint[width * height]] }];
    }
    public static void CheckSize(int width, int height)
    {
        if (width is < 1 or > MaxSide || height is < 1 or > MaxSide)
            throw new InvalidDataException($"Canvas must be 1–{MaxSide} pixels per side.");
    }
    public void Validate()
    {
        CheckSize(Width, Height);
        if (!double.IsFinite(Fps) || Fps is < 1 or > 24 || Layers.Count is < 1 or > MaxLayers)
            throw new InvalidDataException("Use 1–24 FPS and 1–16 layers.");
        if (FrameCount is < 1 or > MaxFrames || Layers.Any(l => !double.IsFinite(l.Opacity) || l.Opacity is < 0 or > 1 ||
                l.Frames.Count != FrameCount || l.Frames.Any(f => f.Length != Width * Height)))
            throw new InvalidDataException("Invalid layer/frame dimensions or opacity.");
    }
    public PixelDocument Clone() => new(Width, Height)
    { Name = Name, Description = Description, Fps = Fps, Layers = Layers.Select(l => l.Clone()).ToList() };
    public bool ContentEquals(PixelDocument other) => Width == other.Width && Height == other.Height && Fps == other.Fps &&
        Name == other.Name && Description == other.Description && Layers.Count == other.Layers.Count &&
        Layers.Zip(other.Layers).All(pair => pair.First.Name == pair.Second.Name && pair.First.Opacity == pair.Second.Opacity &&
            pair.First.Frames.Count == pair.Second.Frames.Count && pair.First.Frames.Zip(pair.Second.Frames).All(f => f.First.AsSpan().SequenceEqual(f.Second)));
    public bool Contains(PixelPoint p) => p.X >= 0 && p.Y >= 0 && p.X < Width && p.Y < Height;

    public uint[] Composite(int frame)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(frame);
        if (frame >= FrameCount) throw new ArgumentOutOfRangeException(nameof(frame));
        var result = new uint[Width * Height];
        foreach (var layer in Layers.Where(l => l.Opacity > 0))
            for (var i = 0; i < result.Length; i++)
            {
                var source = layer.Frames[frame][i];
                if (source >> 24 == 0) continue;
                result[i] = source >> 24 == 255 && layer.Opacity == 1 ? source : Blend(source, result[i], layer.Opacity);
            }
        return result;
    }
    public static uint Blend(uint source, uint dest, double opacity)
    {
        var sa = (source >> 24) / 255d * opacity;
        var da = (dest >> 24) / 255d;
        var a = sa + da * (1 - sa);
        if (a <= 0) return 0;
        uint Channel(int shift)
        {
            var value = (((source >> shift) & 255) * sa + ((dest >> shift) & 255) * da * (1 - sa)) / a;
            return (uint)Math.Clamp(Math.Round(value, MidpointRounding.AwayFromZero), 0, 255);
        }
        return (uint)Math.Round(a * 255, MidpointRounding.AwayFromZero) << 24 | Channel(16) << 16 | Channel(8) << 8 | Channel(0);
    }

    public void Draw(PixelTool tool, PixelPoint start, PixelPoint end, uint color, int brush, int layer, int frame)
    {
        if (!Contains(start) || !Contains(end)) return;
        if (tool == PixelTool.Eyedropper) return;
        var pixels = Layers[layer].Frames[frame];
        if (tool == PixelTool.Fill) { Flood(pixels, end, color); return; }
        var ink = tool == PixelTool.Eraser ? 0 : color;
        var size = Math.Clamp(brush, 1, 8); var offset = (size - 1) / 2;
        void Stamp(int x, int y)
        {
            for (var py = Math.Max(0, y - offset); py < Math.Min(Height, y - offset + size); py++)
                for (var px = Math.Max(0, x - offset); px < Math.Min(Width, x - offset + size); px++)
                    pixels[py * Width + px] = ink;
        }
        var left = Math.Min(start.X, end.X); var right = Math.Max(start.X, end.X);
        var top = Math.Min(start.Y, end.Y); var bottom = Math.Max(start.Y, end.Y);
        if (tool == PixelTool.Rectangle)
        {
            for (var x = left; x <= right; x++) { Stamp(x, top); Stamp(x, bottom); }
            for (var y = top; y <= bottom; y++) { Stamp(left, y); Stamp(right, y); }
        }
        else if (tool == PixelTool.Ellipse && left != right && top != bottom)
        {
            var rx = (right - left) / 2d; var ry = (bottom - top) / 2d; var cx = left + rx; var cy = top + ry;
            for (var x = left; x <= right; x++)
            {
                var dy = ry * Math.Sqrt(Math.Max(0, 1 - Math.Pow((x - cx) / rx, 2)));
                Stamp(x, (int)Math.Round(cy - dy)); Stamp(x, (int)Math.Round(cy + dy));
            }
            for (var y = top; y <= bottom; y++)
            {
                var dx = rx * Math.Sqrt(Math.Max(0, 1 - Math.Pow((y - cy) / ry, 2)));
                Stamp((int)Math.Round(cx - dx), y); Stamp((int)Math.Round(cx + dx), y);
            }
        }
        else
        {
            var x = start.X; var y = start.Y; var dx = Math.Abs(end.X - x); var dy = -Math.Abs(end.Y - y);
            var sx = x < end.X ? 1 : -1; var sy = y < end.Y ? 1 : -1; var error = dx + dy;
            while (true)
            {
                Stamp(x, y); if (x == end.X && y == end.Y) break;
                var twice = error * 2;
                if (twice >= dy) { error += dy; x += sx; }
                if (twice <= dx) { error += dx; y += sy; }
            }
        }
    }
    private void Flood(uint[] pixels, PixelPoint point, uint color)
    {
        var origin = point.Y * Width + point.X; var old = pixels[origin];
        if (old == color) return;
        var queue = new Queue<int>(); queue.Enqueue(origin); pixels[origin] = color;
        void Visit(int index) { if (pixels[index] == old) { pixels[index] = color; queue.Enqueue(index); } }
        while (queue.TryDequeue(out var index))
        {
            var x = index % Width; var y = index / Width;
            if (x > 0) Visit(index - 1); if (x < Width - 1) Visit(index + 1);
            if (y > 0) Visit(index - Width); if (y < Height - 1) Visit(index + Width);
        }
    }
    public void Resize(int width, int height)
    {
        CheckSize(width, height);
        foreach (var layer in Layers)
            for (var frame = 0; frame < FrameCount; frame++)
            {
                var next = new uint[width * height];
                for (var y = 0; y < Math.Min(height, Height); y++)
                    Array.Copy(layer.Frames[frame], y * Width, next, y * width, Math.Min(width, Width));
                layer.Frames[frame] = next;
            }
        Width = width; Height = height;
    }
    public void AddFrame(int after, bool duplicate)
    {
        if (FrameCount >= MaxFrames) return;
        foreach (var layer in Layers) layer.Frames.Insert(after + 1, duplicate ? (uint[])layer.Frames[after].Clone() : new uint[Width * Height]);
    }
    public void DeleteFrame(int index)
    {
        if (FrameCount <= 1) return;
        foreach (var layer in Layers) layer.Frames.RemoveAt(index);
    }
    public void MoveFrame(int from, int to)
    {
        if (to < 0 || to >= FrameCount) return;
        foreach (var layer in Layers) { var frame = layer.Frames[from]; layer.Frames.RemoveAt(from); layer.Frames.Insert(to, frame); }
    }
    public void AddLayer()
    {
        if (Layers.Count >= MaxLayers) return;
        Layers.Add(new() { Name = $"Layer {Layers.Count + 1}", Frames = Enumerable.Range(0, FrameCount).Select(_ => new uint[Width * Height]).ToList() });
    }
    public void Flip(int layer, int frame, bool horizontal)
    {
        var old = Layers[layer].Frames[frame]; var next = new uint[old.Length];
        for (var y = 0; y < Height; y++) for (var x = 0; x < Width; x++)
            next[y * Width + x] = old[(horizontal ? y : Height - y - 1) * Width + (horizontal ? Width - x - 1 : x)];
        Layers[layer].Frames[frame] = next;
    }
}

File: Sources/Unfold/CharacterEditor/PixelDocument.swift
import Foundation

/// Straight-alpha RGBA, in top-left row-major order. No renderer-owned buffers.
struct PixelFrame: Equatable {
    var pixels: [UInt32]

    init(width: Int, height: Int) {
        pixels = Array(repeating: 0, count: width * height)
    }
}

struct PixelLayer: Equatable {
    var name: String
    var opacity: Double = 1
    var frames: [PixelFrame]
}

struct PixelPoint: Equatable {
    let x: Int
    let y: Int
}

enum PixelTool: String, CaseIterable, Identifiable {
    case pencil, eraser, fill, eyedropper, line, rectangle, ellipse
    var id: String { rawValue }
    var title: String { rawValue.capitalized }
    var symbol: String {
        switch self {
        case .pencil: return "pencil.tip"
        case .eraser: return "eraser"
        case .fill: return "drop.fill"
        case .eyedropper: return "eyedropper"
        case .line: return "line.diagonal"
        case .rectangle: return "rectangle"
        case .ellipse: return "circle"
        }
    }
}

struct PixelDocument: Equatable {
    static let maximumLayers = 16
    var width: Int
    var height: Int
    var fps: Double = 12
    var name = "Unfold Character"
    var description = ""
    /// Bottom layer first, matching Piskel's serialized layer order.
    var layers: [PixelLayer]
    var frameCount: Int { layers[0].frames.count }
    var byteCount: Int { width * height * frameCount * layers.count * 4 }

    init(width: Int = Constants.editorDefaultCanvasSide, height: Int = Constants.editorDefaultCanvasSide) {
        self.width = min(max(width, 1), Constants.editorCanvasSideRange.upperBound)
        self.height = min(max(height, 1), Constants.editorCanvasSideRange.upperBound)
        layers = [PixelLayer(name: "Layer 1", frames: [PixelFrame(width: self.width, height: self.height)])]
    }

    func contains(_ point: PixelPoint) -> Bool {
        (0..<width).contains(point.x) && (0..<height).contains(point.y)
    }

    func compositedFrame(at index: Int) -> PixelFrame {
        var result = PixelFrame(width: width, height: height)
        guard (0..<frameCount).contains(index) else { return result }
        for layer in layers where layer.opacity > 0 {
            for i in result.pixels.indices {
                let source = layer.frames[index].pixels[i]
                if source & 255 == 0 { continue }
                result.pixels[i] = source & 255 == 255 && layer.opacity == 1
                    ? source : Self.composite(source, over: result.pixels[i], opacity: layer.opacity)
            }
        }
        return result
    }

    static func composite(_ source: UInt32, over destination: UInt32, opacity: Double) -> UInt32 {
        let sa = Double(source & 255) / 255 * opacity
        let da = Double(destination & 255) / 255
        let alpha = sa + da * (1 - sa)
        guard alpha > 0 else { return 0 }
        var result = UInt32((alpha * 255).rounded())
        for shift in [24, 16, 8] {
            let s = Double((source >> shift) & 255)
            let d = Double((destination >> shift) & 255)
            let channel = UInt32(((s * sa + d * da * (1 - sa)) / alpha).rounded())
            result |= min(channel, 255) << shift
        }
        return result
    }

    mutating func draw(tool: PixelTool, from start: PixelPoint, to end: PixelPoint,
                       color: UInt32, brush: Int, layer: Int, frame: Int) {
        guard layers.indices.contains(layer), (0..<frameCount).contains(frame) else { return }
        if tool == .fill {
            floodFill(at: end, color: color, layer: layer, frame: frame)
            return
        }
        guard tool != .eyedropper else { return }
        let ink: UInt32 = tool == .eraser ? 0 : color
        let size = min(max(brush, 1), 8)
        func stamp(_ point: PixelPoint, into pixels: inout [UInt32]) {
            let offset = (size - 1) / 2
            for y in (point.y - offset)..<(point.y - offset + size) {
                for x in (point.x - offset)..<(point.x - offset + size) where (0..<width).contains(x) && (0..<height).contains(y) {
                    pixels[y * width + x] = ink
                }
            }
        }
        var pixels = layers[layer].frames[frame].pixels
        let left = min(start.x, end.x), right = max(start.x, end.x)
        let top = min(start.y, end.y), bottom = max(start.y, end.y)
        if tool == .rectangle {
            for x in left...right {
                stamp(PixelPoint(x: x, y: top), into: &pixels)
                stamp(PixelPoint(x: x, y: bottom), into: &pixels)
            }
            for y in top...bottom {
                stamp(PixelPoint(x: left, y: y), into: &pixels)
                stamp(PixelPoint(x: right, y: y), into: &pixels)
            }
        } else if tool == .ellipse && left != right && top != bottom {
            let rx = Double(right - left) / 2, ry = Double(bottom - top) / 2
            let cx = Double(left) + rx, cy = Double(top) + ry
            // Sample both axes to keep narrow ellipses connected.
            for x in left...right {
                let dy = ry * sqrt(max(0, 1 - pow((Double(x) - cx) / rx, 2)))
                stamp(PixelPoint(x: x, y: Int((cy - dy).rounded())), into: &pixels)
                stamp(PixelPoint(x: x, y: Int((cy + dy).rounded())), into: &pixels)
            }
            for y in top...bottom {
                let dx = rx * sqrt(max(0, 1 - pow((Double(y) - cy) / ry, 2)))
                stamp(PixelPoint(x: Int((cx - dx).rounded()), y: y), into: &pixels)
                stamp(PixelPoint(x: Int((cx + dx).rounded()), y: y), into: &pixels)
            }
        } else {
            // Bresenham bridges sparse mouse events without gaps.
            var x = start.x, y = start.y
            let dx = abs(end.x - x), dy = -abs(end.y - y)
            let sx = x < end.x ? 1 : -1, sy = y < end.y ? 1 : -1
            var error = dx + dy
            while true {
                stamp(PixelPoint(x: x, y: y), into: &pixels)
                if x == end.x && y == end.y { break }
                let twice = 2 * error
                if twice >= dy { error += dy; x += sx }
                if twice <= dx { error += dx; y += sy }
            }
        }
        layers[layer].frames[frame].pixels = pixels
    }

    private mutating func floodFill(at point: PixelPoint, color: UInt32, layer: Int, frame: Int) {
        guard contains(point) else { return }
        var pixels = layers[layer].frames[frame].pixels
        let origin = point.y * width + point.x
        let old = pixels[origin]
        guard old != color else { return }
        var queue = [origin]
        pixels[origin] = color
        var cursor = 0
        while cursor < queue.count {
            let index = queue[cursor]
            cursor += 1
            let x = index % width, y = index / width
            for p in [PixelPoint(x: x - 1, y: y), PixelPoint(x: x + 1, y: y),
                      PixelPoint(x: x, y: y - 1), PixelPoint(x: x, y: y + 1)] where contains(p) {
                let next = p.y * width + p.x
                if pixels[next] == old {
                    pixels[next] = color
                    queue.append(next)
                }
            }
        }
        layers[layer].frames[frame].pixels = pixels
    }

    mutating func resize(width newWidth: Int, height newHeight: Int) {
        guard Constants.editorCanvasSideRange.contains(newWidth), Constants.editorCanvasSideRange.contains(newHeight) else { return }
        for l in layers.indices {
            for f in 0..<frameCount {
                var resized = PixelFrame(width: newWidth, height: newHeight)
                for y in 0..<min(height, newHeight) {
                    for x in 0..<min(width, newWidth) {
                        resized.pixels[y * newWidth + x] = layers[l].frames[f].pixels[y * width + x]
                    }
                }
                layers[l].frames[f] = resized
            }
        }
        width = newWidth
        height = newHeight
    }

    mutating func insertFrame(after index: Int, duplicate: Bool) {
        guard frameCount < Constants.editorFrameCountRange.upperBound, (0..<frameCount).contains(index) else { return }
        for l in layers.indices {
            let newFrame = duplicate ? layers[l].frames[index] : PixelFrame(width: width, height: height)
            layers[l].frames.insert(newFrame, at: index + 1)
        }
    }

    mutating func removeFrame(at index: Int) {
        guard frameCount > 1, (0..<frameCount).contains(index) else { return }
        for l in layers.indices { layers[l].frames.remove(at: index) }
    }

    mutating func moveFrame(from: Int, to: Int) {
        guard (0..<frameCount).contains(from), (0..<frameCount).contains(to), from != to else { return }
        for l in layers.indices {
            let frame = layers[l].frames.remove(at: from)
            layers[l].frames.insert(frame, at: to)
        }
    }

    mutating func flip(layer: Int, frame: Int, horizontal: Bool) {
        let old = layers[layer].frames[frame].pixels
        for y in 0..<height {
            for x in 0..<width {
                let sx = horizontal ? width - 1 - x : x
                let sy = horizontal ? y : height - 1 - y
                layers[layer].frames[frame].pixels[y * width + x] = old[sy * width + sx]
            }
        }
    }
}
